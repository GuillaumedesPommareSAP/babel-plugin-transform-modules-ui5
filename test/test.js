/* global it, expect */
const fse = require('fs-extra')
const Path = require('path')
const babel = require('babel-core')
const plugin = require('..')

const fixtureDirPath = Path.resolve(__dirname, 'fixtures')
const resultsDir = Path.join(__dirname, 'output')

fse
  .emptyDirSync(resultsDir)

fse
  .readdirSync(fixtureDirPath)
  .filter(item => item.endsWith('.js'))
  .forEach(filename => {
    it(filename, () => {
      console.log(`Running ${filename}`)
      const filepath = Path.join(fixtureDirPath, filename)
      const babelOpts = {
        plugins: [
          'babel-plugin-syntax-decorators',
          'babel-plugin-syntax-class-properties',
          plugin
        ]
      }
      const result = babel.transformFileSync(filepath, babelOpts).code
      const resultFilepath = Path.join(resultsDir, filename)
      fse.writeFileSync(resultFilepath, result)
      expect(result).toMatchSnapshot()
    })
  })
