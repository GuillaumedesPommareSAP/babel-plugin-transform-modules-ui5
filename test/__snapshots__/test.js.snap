// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`classes class-convert-dec-alias.js 1`] = `
"sap.ui.define(['sap/SAPClass'], function (__SAPClass) {
  function _interopRequireDefault(obj) {
    return obj && obj.__esModule ? obj.default : obj;
  }

  const SAPClass = _interopRequireDefault(__SAPClass);

  const MyClass = SAPClass.extend('x.y.Z', {});
  return MyClass;
});"
`;

exports[`classes class-convert-dec-name.js 1`] = `
"sap.ui.define(['sap/SAPClass'], function (__SAPClass) {
  function _interopRequireDefault(obj) {
    return obj && obj.__esModule ? obj.default : obj;
  }

  const SAPClass = _interopRequireDefault(__SAPClass);

  const MyClass = SAPClass.extend('x.y.Z', {});
  return MyClass;
});"
`;

exports[`classes class-convert-dec-namespace.js 1`] = `
"sap.ui.define(['sap/SAPClass'], function (__SAPClass) {
  function _interopRequireDefault(obj) {
    return obj && obj.__esModule ? obj.default : obj;
  }

  const SAPClass = _interopRequireDefault(__SAPClass);

  const MyClass = SAPClass.extend('x.y.MyClass', {});
  return MyClass;
});"
`;

exports[`classes class-convert-dec-nonui5.js 1`] = `
"sap.ui.define(['other/NotSAPClass'], function (__NotSAPClass) {
  function _interopRequireDefault(obj) {
    return obj && obj.__esModule ? obj.default : obj;
  }

  const NotSAPClass = _interopRequireDefault(__NotSAPClass);

  @NONUI5
  class MyClass extends NotSAPClass {}
  return MyClass;
});"
`;

exports[`classes class-convert-file-namespace.js 1`] = `
"sap.ui.define(['sap/SAPClass'], function (__SAPClass) {
  function _interopRequireDefault(obj) {
    return obj && obj.__esModule ? obj.default : obj;
  }

  const SAPClass = _interopRequireDefault(__SAPClass);

  const MyClass = SAPClass.extend('test.fixtures.classes.MyClass', {});
  return MyClass;
});"
`;

exports[`classes class-convert-file-namespace-prefixed.js 1`] = `
"sap.ui.define(['sap/SAPClass'], function (__SAPClass) {
  function _interopRequireDefault(obj) {
    return obj && obj.__esModule ? obj.default : obj;
  }

  const SAPClass = _interopRequireDefault(__SAPClass);

  const MyClass = SAPClass.extend('prefix.test.fixtures.classes.MyClass', {});
  return MyClass;
});"
`;

exports[`classes class-convert-file-namespace-sourceroot.js 1`] = `
"sap.ui.define(['sap/SAPClass'], function (__SAPClass) {
  function _interopRequireDefault(obj) {
    return obj && obj.__esModule ? obj.default : obj;
  }

  const SAPClass = _interopRequireDefault(__SAPClass);

  const MyClass = SAPClass.extend('classes.MyClass', {});
  return MyClass;
});"
`;

exports[`classes class-convert-jsdoc-alias.js 1`] = `
"sap.ui.define(['sap/SAPClass'], function (__SAPClass) {
  function _interopRequireDefault(obj) {
    return obj && obj.__esModule ? obj.default : obj;
  }

  const SAPClass = _interopRequireDefault(__SAPClass);

  const MyClass = SAPClass.extend('x.y.Z', {});
  return MyClass;
});"
`;

exports[`classes class-convert-jsdoc-name.js 1`] = `
"sap.ui.define(['sap/SAPClass'], function (__SAPClass) {
  function _interopRequireDefault(obj) {
    return obj && obj.__esModule ? obj.default : obj;
  }

  const SAPClass = _interopRequireDefault(__SAPClass);

  const MyClass = SAPClass.extend('x.y.Z', {});
  return MyClass;
});"
`;

exports[`classes class-convert-jsdoc-namespace.js 1`] = `
"sap.ui.define(['sap/SAPClass'], function (__SAPClass) {
  function _interopRequireDefault(obj) {
    return obj && obj.__esModule ? obj.default : obj;
  }

  const SAPClass = _interopRequireDefault(__SAPClass);

  const MyClass = SAPClass.extend('x.y.MyClass', {});
  return MyClass;
});"
`;

exports[`classes class-convert-jsdoc-nonui5.js 1`] = `
"sap.ui.define(['sap/SAPClass'], function (__SAPClass) {
  function _interopRequireDefault(obj) {
    return obj && obj.__esModule ? obj.default : obj;
  }

  const SAPClass = _interopRequireDefault(__SAPClass);

  /**
   * @nonui5
   */
  class MyClass extends SAPClass {}
  return MyClass;
});"
`;

exports[`classes class-convert-metadata-jsdoc.js 1`] = `
"sap.ui.define(['sap/SAPClass'], function (__SAPClass) {
  function _interopRequireDefault(obj) {
    return obj && obj.__esModule ? obj.default : obj;
  }

  const SAPClass = _interopRequireDefault(__SAPClass);

  const mdata = {};

  /**
   * @metadata mdata
   */
  const MyClass = SAPClass.extend('test.fixtures.classes.MyClass', {
    metadata: mdata
  });
  return MyClass;
});"
`;

exports[`classes class-convert-metadata-prop.js 1`] = `
"sap.ui.define(['sap/SAPClass'], function (__SAPClass) {
  function _interopRequireDefault(obj) {
    return obj && obj.__esModule ? obj.default : obj;
  }

  const SAPClass = _interopRequireDefault(__SAPClass);

  const mdata = {};

  const MyClass = SAPClass.extend('test.fixtures.classes.MyClass', {
    metadata: mdata
  });
  return MyClass;
});"
`;

exports[`classes class-convert-metadata-static.js 1`] = `
"sap.ui.define(['sap/SAPClass'], function (__SAPClass) {
  function _interopRequireDefault(obj) {
    return obj && obj.__esModule ? obj.default : obj;
  }

  const SAPClass = _interopRequireDefault(__SAPClass);

  const mdata = {};

  const MyClass = SAPClass.extend('test.fixtures.classes.MyClass', {
    metadata: mdata
  });
  return MyClass;
});"
`;

exports[`classes class-convert-no-constructor.js 1`] = `
"/* global sap */
/**
 * Test notes: Even though there is an existing sp.ui.define,
 * the class should be converted.
 */
sap.ui.define(['sap/SAPClass'], SAPClass => {
  /**
   * @name com.app.MyClass
   */
  const MyClass = SAPClass.extend('com.app.MyClass', {
    constructor: function contructor() {
      SAPClass.prototype.contructor.apply(this, arguments);

      this.property_func = () => {
        return this.getProperty('/name');
      };
    }
  });
  return MyClass;
});"
`;

exports[`classes class-convert-with-existing-define-defer-return.js 1`] = `
"/* global sap */
/**
 * Test notes: Even though there is an existing sp.ui.define,
 * the class should be converted.
 */
sap.ui.define(['sap/SAPClass'], SAPClass => {
  /**
   * @UI5
   * @name com.app.MyClass
   * @namespace com.app
   */
  const MyClass = SAPClass.extend('com.app.MyClass', {
    metadata: {
      thing: \\"string\\"
      // This incorrect shorthand will get fixed
    },
    renderer: {},
    constructor: function (data) {
      SAPClass.constructor.apply(this, [data]);

      this.property_func = () => {
        return this.getProperty('/name');
      };
    }
  });

  return MyClass;
});"
`;

exports[`classes class-convert-with-existing-define-immediate-return.js 1`] = `
"/* global sap */
/**
 * Test notes: Even though there is an existing sp.ui.define,
 * the class should be converted.
 */
sap.ui.define(['sap/SAPClass'], SAPClass => {
  /**
   * @UI5
   * @name com.app.MyClass
   * @namespace com.app
   */
  const MyClass = SAPClass.extend('com.app.MyClass', {
    metadata: {
      thing: \\"string\\"
      // This incorrect shorthand will get fixed
    },
    renderer: {},
    constructor: function (data) {
      SAPClass.constructor.apply(this, [data]);

      this.property_func = () => {
        return this.getProperty('/name');
      };
    }
  });
  return MyClass;
});"
`;

exports[`classes class-covert-nonui5-return.js 1`] = `
"/**
 * This fixture tests a nonui5 scenario inside an sap.ui.define.
 */
sap.ui.define(['sap/SAPClass'], SAPClass => {
  /**
   * @nonui5
   */
  return class MyClass extends SAPClass {};
});"
`;

exports[`classes class-extend-managed.js 1`] = `
"sap.ui.define(['sap/ui/core/Controller', './Relative'], function (__Controller, __Relative) {
  function _interopRequireDefault(obj) {
    return obj && obj.__esModule ? obj.default : obj;
  }

  const Controller = _interopRequireDefault(__Controller);

  const Relative = _interopRequireDefault(__Relative);

  const MyController = Controller.extend('test.fixtures.classes.MyController', {
    renderer: {},
    metadata: {},
    constructor: function (arg) {
      console.log(\\"constructor\\");
      Controller.constructor.apply(this, [arg]);

      this.property_arrow_function = () => 1;
    },
    shorthand_function: function (...args) {
      Controller.prototype.foo.apply(this, [...args]);
      this.name = Relative.name();
    },
    async_shorthand_function: async function (arg) {
      return await 1;
    },
    property_value: 1
  });

  MyController.static_shorthand_function = function () {
    return 1;
  };

  MyController.static_property_value = 1;

  MyController.static_property_arrow = () => 1;

  MyController.static_async_shorthand_function = async function () {
    return await 1;
  };

  MyController.z = function () {
    return 'hey';
  };
  return MyController;
});"
`;

exports[`classes class-fix-with-existing-define.js 1`] = `
"/* global sap */
/**
 * Test notes: Even though there is an existing sp.ui.define,
 * and it uses .extend() syntax, the constructor function should be fixed.
 */
sap.ui.define(['sap/SAPClass'], SAPClass => {
  const X = SAPClass.extend('X', {
    // This incorrect shorthand will get fixed
    constructor: function constructor(data) {
      SAPClass.prototype.constructor.call(this, data);
    },

    method() {
      return this.getProperty('/name');
    }
  });
  return X;
});"
`;

exports[`examples Animal.js 1`] = `
"sap.ui.define([\\"sap/ui/base/ManagedObject\\"], function (__ManagedObject) {
    function _interopRequireDefault(obj) {
        return obj && obj.__esModule ? obj.default : obj;
    }

    const ManagedObject = _interopRequireDefault(__ManagedObject);

    const Animal = ManagedObject.extend(\\"test.fixtures.examples.Animal\\", {
        metadata: {
            properties: {
                type: { type: \\"string\\" },
                nickName: { type: \\"string\\" }
            }
        },
        constructor: function (...args) {
            ManagedObject.constructor.apply(this, [...args]);
        },
        init: function () {
            ManagedObject.prototype.init.apply(this, []);
        },
        callMe: function () {
            alert(\`I'm a \${this.getType()}.
        Call me \${this.getNickName()}.\`);
        }
    });
    return Animal;
});"
`;

exports[`examples Cat.js 1`] = `
"sap.ui.define([\\"./Animal\\"], function (__Animal) {
    function _interopRequireDefault(obj) {
        return obj && obj.__esModule ? obj.default : obj;
    }

    const Animal = _interopRequireDefault(__Animal);

    const Cat = Animal.extend(\\"othernamespace.Cat\\", {
        init: function () {
            Animal.prototype.init.apply(this, []);
            this.setType(\\"Cat\\");
        },
        callMe: function () {
            Animal.prototype.callMe.apply(this, []);
            alert(\\"Miao~\\");
        }
    });

    Cat.createCat = function (nickName) {
        const cat = new example.obj.Cat({
            nickName
        });
        return cat;
    };

    return Cat;
});"
`;

exports[`examples MyError.js 1`] = `
"sap.ui.define([], function () {
  class MyError extends Error {
    constructor(msg) {
      super(msg);
      this.name = 'MyError';
    }
  }
  return MyError;
});"
`;

exports[`examples Utils.js 1`] = `
"sap.ui.define([], function () {
  const exports = {};
  Object.defineProperty(exports, \\"__esModule\\", {
    value: true
  });
  function multiply(a, b) {
    return a * b;
  }

  function add(a, b) {
    // Not a named export but gets added for sap.ui.define() interop.
    return a + b;
  }

  exports.multiply = multiply;
  exports.default = {
    multiply, add
  };
  exports.add = add;
  return exports;
});"
`;

exports[`exports export-class.js 1`] = `
"sap.ui.define([], function () {
  function hello() {
    return 'hey';
  }

  class Thing {
    constructor() {
      console.log('constructor');
    }
  }

  Thing.foo = hello;
  return Thing;
});"
`;

exports[`exports export-class-extend-native.js 1`] = `
"sap.ui.define([], function () {
  class MyError extends Error {
    constructor(m) {
      super(m);
      console.log('constructor');
      this.name = 'MyError';
    }
  }
  return MyError;
});"
`;

exports[`exports export-default-anon-function.js 1`] = `
"sap.ui.define([], function () {
  let one = 1;

  let two = 2;
  return async function () {
    // Since it's anonymous, this can be safetly moved to the end
    return (await one) + two;
  };
});"
`;

exports[`exports export-default-arrow-function.js 1`] = `
"sap.ui.define([], function () {
  let one = 1;

  let two = 2;
  return () => {
    return one + two;
  };
});"
`;

exports[`exports export-default-identifier.js 1`] = `
"sap.ui.define([], function () {
  const x = {};

  x.a = 1;
  return x;
});"
`;

exports[`exports export-default-named-function.js 1`] = `
"sap.ui.define([], function () {
  let one = 1;

  function add() {
    return one + two + this.x;
  }

  add.x = 3;

  let two = 2;
  return add;
});"
`;

exports[`exports export-interop.js 1`] = `
"sap.ui.define([], function () {
  const exports = {};
  Object.defineProperty(exports, \\"__esModule\\", {
    value: true
  });

  // This tests the scenario when there is a mix of named and default exports.
  // If the importing module has an interop, there's no issue with that.
  // But if not, the importing module will get the 'exports' object and not the 'default' object.
  // If the 'exports' doesn't have all the properties that 'default' has, the importing module will get an unexpected undefined.


  function notExported() {}

  function exported() {}

  function bar() {}

  exports.exported = exported;
  exports.bar = bar;
  exports.default = {
    notExported,
    exported,
    baz: bar,
    one: 1
  };
  exports.notExported = notExported;
  exports.baz = bar;
  exports.one = 1;
  return exports;
});"
`;

exports[`exports export-interop-error-obj-method.js 1`] = `": Can't build interop export since default export's key 'fn' conflicts with named export."`;

exports[`exports export-interop-error-obj-prop.js 1`] = `": Can't build interop export since default export's key 'a' conflicts with named export."`;

exports[`exports export-multiple.js 1`] = `
"sap.ui.define([], function () {
  const exports = {};
  Object.defineProperty(exports, \\"__esModule\\", {
    value: true
  });
  function def() {
    console.log(\\"I'm the default\\");
  }

  const thing1 = {};
  const thing2 = 2;
  const thingy3 = 3;

  const c = 1;
  var v = 1;
  let l = 1;
  function f() {}

  let laterLet;
  laterLet = \\"now\\";

  var laterVar;
  laterVar = \\"var\\";
  exports.def = def;
  exports.thing1 = thing1;
  exports.thing2 = thing2;
  exports.thing3 = thingy3;
  exports.c = c;
  exports.v = v;
  exports.l = l;
  exports.f = f;
  exports.laterLet = laterLet;
  exports.laterVar = laterVar;
  exports.default = def;
  return exports;
});"
`;

exports[`exports export-named-as-default.js 1`] = `
"sap.ui.define([], function () {
  const exports = {};
  Object.defineProperty(exports, \\"__esModule\\", {
    value: true
  });
  const one = 1;
  exports.default = one;
  return exports;
});"
`;

exports[`exports export-obj.js 1`] = `
"sap.ui.define([], function () {
  // Hello World

  function getFoo() {
    return 1;
  }

  function bar() {
    return 2;
  }

  return {
    foo: getFoo(),
    bar
  };
});"
`;

exports[`imports import-multiple.js 1`] = `
"sap.ui.define(['./../a/b', 'A', 'BC', 'DEF', 'G', 'HI'], function (____a_b, __A, __BC, __D, __G, __H) {
  function _interopRequireDefault(obj) {
    return obj && obj.__esModule ? obj.default : obj;
  }

  const A = _interopRequireDefault(__A);

  const B = __BC['B'];
  const C = __BC['C'];

  const D = _interopRequireDefault(__D);

  const E = __D['E'];
  const F = __D['F'];
  const G = __G;

  const H = _interopRequireDefault(__H);

  const I = __H;
});"
`;

exports[`libs libs-sap-imports.js 1`] = `
"sap.ui.define(['sap/SAPClass'], function (__SAPClass) {
  function _interopRequireDefault(obj) {
    return obj && obj.__esModule ? obj.default : obj;
  }

  const SAPClass = _interopRequireDefault(__SAPClass);

  const Levels = __SAPClass['Levels'];
  const MyClass = SAPClass.extend('test.fixtures.libs.MyClass', {});
  MyClass.Levels = Levels;
  return MyClass;
});"
`;

exports[`mixed mixed-import-export.js 1`] = `
"sap.ui.define(['1', '2', '3', '4', '5'], function (__1, __2, __3, __4, __5) {
  const exports = {};
  Object.defineProperty(exports, \\"__esModule\\", {
    value: true
  });

  function extendExports(exports, obj) {
    Object.keys(obj).forEach(function (key) {
      if (key === \\"default\\" || key === \\"__esModule\\") return;
      Object.defineProperty(exports, key, {
        enumerable: true,
        get: function get() {
          return obj[key];
        }
      });
    });
  }

  extendExports(exports, __1);
  exports.a = __2.a;
  exports.b2 = __2.b;
  exports.default = __3.default;
  exports.default2 = __4.default;
  exports.x = __5.v;
  return exports;
});"
`;

exports[`no-wrap nowrap-existing-sap.ui.define.js 1`] = `
"/* global sap */
sap.ui.define(['JSONModel'], JSONModel => {
  return new JSONModel();
});"
`;

exports[`no-wrap skip-iffe.js 1`] = `
"(function () {
  window.x = {};
})();"
`;

exports[`wrapping wrapping-export.js 1`] = `
"sap.ui.define([], function () {
  return {};
}, true);"
`;

exports[`wrapping wrapping-export-const.js 1`] = `
"sap.ui.define([], function () {
  const X = {};

  /**
   * @export
   */
  return X;
}, true);"
`;
